man/bFst.1
.\" Automatically generated by Pandoc 2.19.2
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "BFST" "1" "" "bFst (vcflib)" "bFst (VCF statistics)"
.hy
.SH NAME
.PP
\f[B]bFst\f[R]
.SH SYNOPSIS
.PP
\f[B]bFst\f[R] \[en]target 0,1,2,3,4,5,6,7 \[en]background
11,12,13,16,17,19,22 \[en]file my.vcf \[en]deltaaf 0.1
.SH DESCRIPTION
.PP
\f[B]bFst\f[R] is a Bayesian approach to Fst.
Importantly \f[B]bFst\f[R] accounts for genotype uncertainty in the
model using genotype likelihoods.
For a more detailed description see: \[ga]A Bayesian approach to
inferring population structure from dominant markers\[cq] by Holsinger
et al.\ Molecular Ecology Vol 11, issue 7 2002.
The likelihood function has been modified to use genotype likelihoods
provided by variant callers.
There are five free parameters estimated in the model: each
subpopulation\[cq]s allele frequency and Fis (fixation index, within
each subpopulation), a free parameter for the total population\[cq]s
allele frequency, and Fst.
.SH OPTIONS
.IP
.nf
\f[C]


Output : 11 columns :                          
     1.  Seqid                                     
     2.  Position                    
     3.  Observed allele frequency in target.        
     4.  Estimated allele frequency in target.     
     5.  Observed allele frequency in background.  
     6.  Estimated allele frequency in background. 
     7.  Observed allele frequency combined.         
     8.  Estimated allele frequency in combined.   
     9.  ML estimate of Fst (mean)           
     10. Lower bound of the 95% credible interval  
     11. Upper bound of the 95% credible interval  

required: t,target     -- a zero bases comma separated list of target individuals corrisponding to VCF columns
required: b,background -- a zero bases comma separated list of background individuals corrisponding to VCF columns
required: f,file a     -- a proper formatted VCF file.  the FORMAT field MUST contain \[dq]PL\[dq]
required: d,deltaaf    -- skip sites were the difference in allele frequency is less than deltaaf


Type: statistics
\f[R]
.fi
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]not 0\f[R]
Failure
.SH SEE ALSO
.PP
\f[B]vcflib\f[R](1)
.SH OTHER
.SH LICENSE
.PP
Copyright 2011-2026 (C) Erik Garrison and vcflib contributors.
MIT licensed.
Copyright 2020-2026 (C) Pjotr Prins.
.SH AUTHORS
Erik Garrison and vcflib contributors.


RELEASE_NOTES.md
For contributions see
[contributors](https://github.com/vcflib/vcflib/graphs/contributors)
and
[commits](https://github.com/vcflib/vcflib/commits/master).

## ChangeLog ongoing

* Added guix-container-manifest.scm that allows building Singularity and Docker containers(!)

## ChangeLog v1.0.15 (20260319)

* Fix iterator usage in vcfkeepinfo and vcfremoveaberrantgenotypes (Jukka Matilainen)
* Fixes for vcfwave (Glenn Hickey)
* Added include files to build with vg
* Added pkgconfig support
* Fix wfa2lib builds and sent upstream
* Update zig build to 0.15, or later, and fix static build with zig for vcfcreatemulti
* Create tar ball with static binaries
* Fix building with git submodules only
* Download singularity and docker containers from https://files.genenetwork.org/software/vcflib/v1.0.15/
  md5sum:
  5768540e7bef2db1363e422bc37614c4  vcflib-1.0.15-docker.tar.gz # Docker
  680cd7675325561636e03274e679d195  vcflib-1.0.15-singularity.pack.gz # Singularity
  7f60bb29dfb498c1059a33e10e99ac02  vcflib-static-git-v1.0.15-2-g0d40ff1.tar.gz # Static binaries

## ChangeLog v1.0.14 (20250518)

* CMake: Fixed building static libs and shared libs (the previous version just compiled a static lib)
* Updated zig compile for 0.14 (MichaelK)
* Lots of refactoring (MichaelK)
* CMake: Improved linker speeds by moving some libs
* CMake: Improved markdown/pydoc tests, so we don't show the path to the binary
* guix.scm has multiple outputs including static binaries and a test mode
* guix.scm has procedure for building with local htslib and tabixpp only
* Added @RedPenguin100 Michael Kovaliov as a maintainer (welcome!)

## ChangeLog v1.0.13 (20250417)

* Zig support was added to Guix and Conda - and we expect it to be more common soon
* Addind release protocol to README - so I don't forget bumping version etc.
* Fixed vcfwave issues (see https://github.com/vcflib/vcflib/issues/416) and conda build (thanks Michael Kovaliov, aka @RedPenguin100)
* Fixed github actions (Michael)
* Lots of minor refactoring and cleanup, as well as a memory leak (Michael)
* tabix.hpp fixed and set `WFA_INCLUDE_DIRS` and `-DBUILD_SHARED_LIBS=ON` (thanks Michael Cho, aka @cho-m)
* Cmake fixes (thanks Adam Novak)
* Updated simde submodule
* Fix CMakeLists.txt not to try to install wfa-lib when it is a package
* Reverted on Canocalize extraction in Variant class because of downstream use, see https://github.com/vcflib/vcflib/issues/417

## ChangeLog v1.0.12 (20241122)

* Improved parsing of INFO and FORMAT lines by @jeizenga in https://github.com/vcflib/vcflib/pull/374
* Support external tabixpp by @adamnovak in https://github.com/vcflib/vcflib/pull/375
* Fix vcfwave.cpp by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/407
* Use regex to find the '..' between postions and replace it with '-' by @debbyku in https://github.com/vcflib/vcflib/pull/405
* Fix tag-fail long option that was overriden by tag-pass in vcffilter.cpp by @Gullumluvl in https://github.com/vcflib/vcflib/pull/404
* `vcfwave`: fix condition to avoid nullifying valid SNPs and MNPs by @AndreaGuarracino in https://github.com/vcflib/vcflib/pull/408
* Upgraded Zig support to 0.13.0 by @pjotrp
* Merged multichoose code into vcflib since no one else uses it
* Added sources for canonicalize too
* Moved Fasta.h (fastahack) dependencies from Variant.h into sources
* Updated multichoose and simde modules
* Improved vcfwave support

### New Contributors
* @jeizenga made their first contribution in https://github.com/vcflib/vcflib/pull/374
* @debbyku made their first contribution in https://github.com/vcflib/vcflib/pull/405
* @Gullumluvl made their first contribution in https://github.com/vcflib/vcflib/pull/404

## ChangeLog v1.0.11-pre ()

+ Stopped vendoring wfa2lib by default - so now the Debian build command is
  `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DZIG=OFF ..`


## ChangeLog v1.0.10 (20240420)

Vcflib maintenance release

+ Fixed vcfwave bugs - thanks Andrea Guarracino!
+ Fixed c++17 compatability - thanks Alex Petty!
+ Fixed tabixpp script - thanks Blaise Li!
+ Improved for vcfwave the wfa2lib build, fixed running tests with upstream build update
+ Removed deprecated binary_function
+ Removed unused files from repo
+ Updated README

## ChangeLog v1.0.9 (20230211)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.8 (20230210)

Vcflib maintenance release - mostly for including in Debian

+ Another fix so downstream packages, such as freebayes, no longer need WFAlib.

## ChangeLog v1.0.7 (20230207)

Vcflib maintenance release - mostly for including in Debian

+ Fixed regression discovered by garguantua_kerr and atille of Debian (thanks!)
+ Added note on bio-vcf in vcffilter doc
+ notes on vcfcreatemulti and backtracking
+ CMake: honour include(GNUInstallDirs) paths (I forgot)

## ChangeLog v1.0.6 (20230129)

Vcflib maintenance release - mostly for including in Debian

+ Fixed zig complaining about leaking memory
+ Added CMake Debian support with -DWFA_GITMODULE=OFF
+ Introduced CMake include(GNUInstallDirs)
+ Successfully built wfa2 using embedded CMakeLists.txt
+ Cleaned up CMakeLists.txt removing comments etc.
+ Reintroduced vcfcreatemulti in legacy mode when ZIG=OFF (for Debian)

## ChangeLog v1.0.5 (20230116)

Vcflib's first *Humpty Dumpty* release: [vcfcreatemulti](./doc/vcfcreatemulti.md) is the natural companion to [vcfwave](./doc/vcfwave.md).

Often variant callers are not perfect.
**vcfwave** with its companion tool **vcfcreatemulti** can take an existing VCF file that contains multiple complex overlapping and even nested alleles and, unlike Humpty Dumpty, take them apart and put them together again.
Thereby, hopefully, creating sane VCF output that is useful for analysis and getting rid of false positives.

We created these tools by including the state-of-the-art [biWFA](https://github.com/smarco/WFA2-lib) wavefront aligner.

src/bFst.cpp
/*
    vcflib C++ library for parsing and manipulating VCF files

    Copyright © 2010-2020 Erik Garrison
    Copyright © 2020      Pjotr Prins

    This software is published under the MIT License. See the LICENSE file.
*/

#include "Variant.h"
#include "pdflib.hpp"
#include "index.hpp"

#include <string>
#include <iostream>
#include <cmath>
#include <ctime>
#include <getopt.h>
#include "gpatInfo.hpp"

using namespace std;
using namespace vcflib;

struct pop{

  double nalt ;
  double nref ;
  double af   ;
  double nhomr;
  double nhoma;
  double nhet ;
  double ngeno;
  double fis  ;
  vector<int>    questionable;
  vector<int>    geno_index ;
  vector< vector< double > > unphred_p;

};

double unphred(const string& phred){
  double unphred = atof(phred.c_str());
  unphred = unphred / -10;
  return unphred;
}

void initPop(pop & population){
  population.nalt  = 0;
  population.nref  = 0;
  population.af    = 0;
  population.nhomr = 0;
  population.nhoma = 0;
  population.nhet  = 0;
  population.ngeno = 0;
  population.fis   = 0;

}

void loadPop( vector< map< string, vector<string> > >& group, pop & population){
  int index = 0;

  for(auto& target : group){

    string genotype = target["GT"].front();

    vector<double> phreds;

    phreds.push_back( unphred(target["PL"][0]));
    phreds.push_back( unphred(target["PL"][1]));
    phreds.push_back( unphred(target["PL"][2]));

    double scaled ;
    double norm   = log(exp(phreds[0]) + exp(phreds[1]) + exp(phreds[2]));

    population.unphred_p.push_back(phreds);

    while(1){
      if(genotype == "0/0"){
	population.ngeno += 1;
	population.nhomr += 1;
	population.nref  += 2;
	population.geno_index.push_back(0);
	scaled = exp(phreds[0] - norm);
	break;
      }
      if(genotype == "0/1"){
	population.ngeno += 1;
	population.nhet  += 1;
	population.nref  += 1;
	population.nalt  += 1;
	population.geno_index.push_back(1);
	scaled = exp(phreds[1] - norm);
	break;
      }
      if(genotype == "1/1"){
	population.ngeno += 1;
	population.nhoma += 1;
	population.nalt  += 2;
	population.geno_index.push_back(2);
	scaled = exp(phreds[2] - norm);
	break;
      }
      if(genotype == "0|0"){
	population.ngeno += 1;
	population.nhomr += 1;
	population.nref  += 2;
	population.geno_index.push_back(0);
	scaled = exp(phreds[0] - norm);
	break;
      }
      if(genotype == "0|1"){
	population.ngeno += 1;
	population.nhet  += 1;
	population.nref  += 1;
	population.nalt  += 1;
	population.geno_index.push_back(1);
	scaled = exp(phreds[1] - norm);
	break;
      }
      if(genotype == "1|1"){
	population.ngeno += 1;
	population.nhoma += 1;
	population.nalt  += 2;
	population.geno_index.push_back(2);
	scaled = exp(phreds[2] - norm);
	break;
      }
      cerr << "FATAL: unknown genotype" << endl;
      exit(1);
    }

    if(scaled < 0.75){
      population.questionable.push_back(index);
    }
    index += 1;
  }

  if(population.nalt == 0 && population.nref == 0){
    population.af = -1;
  }
  else{

    population.af  = (population.nalt / (population.nref + population.nalt));

    if(population.nhet > 0){
      population.fis = ( 1 - ((population.nhet/population.ngeno) / (2*population.af*(1 - population.af))));
    }
    else{
      population.fis = 1;
    }
    if(population.fis < 0){
      population.fis = 0.00001;
    }
  }
}

double bound(double v){
  if(v <= 0.00001){
    return  0.00001;
  }
  if(v >= 0.99999){
    return 0.99999;
  }
  return v;
}


void phardy(vector<double>& results, double af, double fis){

  double p0 = pow((1 - af),2) + ((1 - af)*af*fis);
  double p1 = 2*(1 - af)*af*(1 - fis);
  double p2 = pow(af,2) + ((1 - af)*af*fis);

  results.push_back(p0);
  results.push_back(p1);
  results.push_back(p2);

}

double likelihood(pop & population, double af, double fis){

  af  = bound(af);
  fis = bound(fis);

  double het;
  double all;
  double ref;
  double alt;
  double loglikelihood = 0;

  vector<double> genotypeProbs;

  phardy(genotypeProbs, af, fis);

  int geno_indx = 0;

  for(const auto idx : population.geno_index){

    double aa = population.unphred_p[geno_indx][0] + log(genotypeProbs[0]);
    double ab = population.unphred_p[geno_indx][1] + log(genotypeProbs[1]);
    double bb = population.unphred_p[geno_indx][2] + log(genotypeProbs[2]);
    double norm = exp(aa) + exp(ab) + exp(bb);

    double prop = population.unphred_p[geno_indx][idx] +  log(genotypeProbs[idx]);

    loglikelihood += (prop - norm);

    geno_indx++;
  }

  return loglikelihood;

}

double FullProb(pop & target, pop & back, vector<double>& p)
{

// parameters targetAf backgroundAf targetFis backgroundFis totalAf fst

  double alpha = ( (1-p[5])/p[5] ) * p[4];
  double beta  = ( (1-p[5])/p[5] ) * (1 - p[4]);


  double afprior  = log( r8_normal_pdf (p[6], 0.1, p[4]));
  double afpriorT = log( r8_normal_pdf (target.af, 0.05, p[0]));
  double afpriorB = log( r8_normal_pdf (back.af,   0.05, p[1]));

  if(std::isinf(afprior) || std::isnan(afprior)){
    return -100000;
  }

  double ptaf = log( r8_beta_pdf(alpha, beta, p[0]) );
  double pbaf = log( r8_beta_pdf(alpha, beta, p[1]) );

  if( std::isinf(ptaf) || std::isnan(ptaf) || std::isinf(pbaf) || std::isnan(pbaf) ){
    return -100000;
  }


  double llt  = likelihood(target, p[0], p[2]);
  double llb  = likelihood(back,   p[1], p[3]);
  double full = llt + llb + ptaf + pbaf + afprior + afpriorT + afpriorB;


  return full;

}



void updateParameters(pop & target, pop & background, vector<double>& parameters, int pindx){

  // parameters targetAf backgroundAf targetFis backgroundFis totalAf fst

  double origpar = parameters[pindx];

  double accept  = ((double)rand() / (double)(RAND_MAX));
  double up      = ((double)rand() / (double)(RAND_MAX))/10 - 0.05;
  double updatep = parameters[pindx] + up;

  //  cerr << accept << "\t" << up << endl;

  if(updatep >= 1 || updatep <= 0){
    return;
  }

  double llB = FullProb(target, background, parameters);
  parameters[pindx] = updatep;
  double llT = FullProb(target, background, parameters);

  if((llT - llB) > accept){
    return;
  }
  else{
    parameters[pindx] = origpar;
  }
}

void updateGenotypes(pop & target, pop & background, vector<double>& parameters, int gindex, int tbindex){

  // tbindex indicates if the subroutine will update the target or background genotype;

  double accept  = ((double)rand() / (double)(RAND_MAX));
  int  newGindex = rand() % 3;

  //cerr << newGindex << endl;
  //cerr << "gindex "   << gindex << endl;
  //cerr << "gsize t:" << target.geno_index.size() << endl;
  //cerr << "gsize b:" << background.geno_index.size() << endl;



  int oldtindex = target.geno_index[gindex] ;
  int oldbindex = background.geno_index[gindex] ;

  double llB = FullProb(target, background, parameters);

  if(tbindex == 0){
    //udate target
    target.geno_index[gindex] = newGindex;
      }
  else{
    // update background
    background.geno_index[gindex] = newGindex;
  }
  double llT = FullProb(target, background, parameters);

  if((llT - llB) > accept){
    return;
  }
  else{
    if(tbindex == 0){
      target.geno_index[gindex] = oldtindex;
    }
    else{
      target.geno_index[gindex] = oldbindex;
    }
  }
}


int cmp(const void *x, const void *y)
{
  double xx = *(double*)x, yy = *(double*)y;
  if (xx < yy) return -1;
  if (xx > yy) return  1;
  return 0;
}


int main(int argc, char** argv) {

  // set the random seed for MCMC

  srand((unsigned)time(NULL));

  // the filename

  string filename = "NA";

  // using vcflib; thanks to Erik Garrison

  VariantCallFile variantFile ;

  // zero based index for the target and background indivudals

  map<int, int> it, ib;

  // deltaaf is the difference of allele frequency we bother to look at

  string deltaaf ;
  double daf  = -1;

    const struct option longopts[] =
      {
	{"version"   , 0, 0, 'v'},
	{"help"      , 0, 0, 'h'},
        {"file"      , 1, 0, 'f'},
	{"target"    , 1, 0, 't'},
	{"background", 1, 0, 'b'},
	{"deltaaf"   , 1, 0, 'd'},
	{0,0,0,0}
      };

    int index;
    int iarg = 0;

    while(iarg != -1)
      {
	iarg = getopt_long(argc, argv, "d:t:b:f:hv", longopts, &index);

	switch (iarg)
	  {
	  case 0:
	    break;
	  case 'h':
	    cerr << endl;

	    cerr << "INFO: help" << endl << endl;

	    cerr << "     bFst is a Bayesian approach to Fst.  Importantly bFst accounts for genotype uncertainty in the model using genotype likelihoods."       << endl;
	    cerr << "     For a more detailed description see: `A Bayesian approach to inferring population structure from dominant markers' by Holsinger et al. Molecular Ecology Vol 11, issue 7 2002.  The likelihood function has been "	     << endl;
	    cerr << "     modified to use genotype likelihoods provided by variant callers. There are five free parameters estimated in the model: each "	     << endl;
	    cerr << "     subpopulation's allele frequency and Fis (fixation index, within each subpopulation), a free parameter for the total population\'s "  << endl;
	    cerr << "     allele frequency, and Fst. "                                                                                      << endl             << endl;

	      cerr << "Output : 11 columns :                          " << endl;
	      cerr << "     1.  Seqid                                     " << endl;
	      cerr << "     2.  Position				     " << endl;
	      cerr << "     3.  Observed allele frequency in target.	     " << endl;
	      cerr << "     4.  Estimated allele frequency in target.     " << endl;
	      cerr << "     5.  Observed allele frequency in background.  " << endl;
	      cerr << "     6.  Estimated allele frequency in background. " << endl;
	      cerr << "     7.  Observed allele frequency combined. 	     " << endl;
	      cerr << "     8.  Estimated allele frequency in combined.   " << endl;
	      cerr << "     9.  ML estimate of Fst (mean)		     " << endl;
	      cerr << "     10. Lower bound of the 95% credible interval  " << endl;
	      cerr << "     11. Upper bound of the 95% credible interval  " << endl << endl;


	    cerr << "INFO: usage:  bFst --target 0,1,2,3,4,5,6,7 --background 11,12,13,16,17,19,22 --file my.vcf --deltaaf 0.1" << endl;
	    cerr << endl;
	    cerr << "INFO: required: t,target     -- a zero bases comma separated list of target individuals corrisponding to VCF columns" << endl;
	    cerr << "INFO: required: b,background -- a zero bases comma separated list of background individuals corrisponding to VCF columns" << endl;
	    cerr << "INFO: required: f,file a     -- a proper formatted VCF file.  the FORMAT field MUST contain \"PL\"" << endl;
	    cerr << "INFO: required: d,deltaaf    -- skip sites were the difference in allele frequency is less than deltaaf" << endl;
	    cerr << endl << endl << "Type: statistics" << endl;
	    printVersion();
	    cerr << endl << endl;
	    return 0;

	  case 'v':
	    printVersion();
	    return 0;

	  case 't':
	    loadIndices(ib, optarg);
	    cerr << "INFO: There are " << ib.size() << " individuals in the target" << endl;
	    break;

	  case 'b':
	    loadIndices(it, optarg);
	    cerr << "INFO: There are " << it.size() << " individuals in the background" << endl;
	    break;

	  case 'f':
	    cerr << "INFO: File: " << optarg  <<  endl;
	    filename = optarg;
	    break;

	  case 'd':
	    cerr << "INFO: difference in allele frequency : " << optarg << endl;
	    deltaaf = optarg;
	    daf = atof(deltaaf.c_str());
	    break;
	  default:
	    break;
	    cerr << endl;
	    cerr << "FATAL: unknown command line option " << optarg << endl << endl ;
	    cerr << "INFO:  please use bFst --help      " << endl;
	    cerr << endl;
	    return(1);
	  }

      }

    if(daf == -1){
    cerr << endl;
      cerr << "FATAL: did not specify deltaaf" << endl;
      cerr << "INFO:  please use bFst --help      " << endl;
      cerr << endl;
      return(1);
    }

    if(filename == "NA"){
      cerr << endl;
      cerr << "FATAL: did not specify VCF file" << endl;
      cerr << "INFO:  please use bFst --help      " << endl;
      cerr << endl;
      return(1);
    }

    variantFile.open(filename);


    if (!variantFile.is_open()) {
      cerr << endl;
      cerr << "FATAL: could not open VCF file" << endl;
      cerr << "INFO:  please use bFst --help" << endl;
      cerr << endl;
      return(1);
    }
    if(it.size() < 2){
      cerr << endl;
      cerr << "FATAL: target not specified or less than two indviduals" << endl;
      cerr << "INFO:  please use bFst --help                          " << endl;
      cerr << endl;
    }
    if(ib.size() < 2){
      cerr << endl;
      cerr << "FATAL: target not specified or less than two indviduals"<< endl;
      cerr << "INFO:  please use bFst --help                          " << endl;
      cerr << endl;
    }

    Variant var(variantFile);

    vector<string> samples = variantFile.sampleNames;
    int nsamples = samples.size();

    while (variantFile.getNextVariant(var)) {

	// biallelic sites naturally

	if(var.alt.size() > 1){
	  continue;
	}


	vector < map< string, vector<string> > > target, background, total;

	int index = 0;

	for(int nsamp = 0; nsamp < nsamples; nsamp++){

          map<string, vector<string> > sample = var.samples[ samples[nsamp]];

	  if(sample["GT"].front() != "./."){
	    if(it.find(index) != it.end() ){
	      target.push_back(sample);
	      total.push_back(sample);

	    }
	    if(ib.find(index) != ib.end()){
		background.push_back(sample);
		total.push_back(sample);
	    }
	  }

	  index += 1;
	}

	if(target.size() < 2 || background.size() < 2 ){
	  continue;
	}

	pop popt, popb, popTotal;

	initPop(popt);
	initPop(popb);
	initPop(popTotal);

	loadPop(target,     popt);
	loadPop(background, popb);
	loadPop(total,  popTotal);

	if(popt.af == -1 || popb.af == -1){
	  continue;
	}
	if(popt.af == 1  && popb.af == 1){
	  continue;
	}
	if(popt.af == 0 && popb.af  == 0){
	  continue;
	}

	double afdiff = abs(popt.af - popb.af);

	if(afdiff < daf){
	  continue;
	}


	cerr << "INFO: target has "     << popt.questionable.size() << " questionable genotypes " << endl;
	cerr << "INFO: background has " << popb.questionable.size() << " questionable genotypes " << endl;

	// Parameters- targetAf backgroundAf targetFis backgroundFis totalAf fst
	vector<double> parameters;
	parameters.push_back(popt.af);
	parameters.push_back(popb.af);
	parameters.push_back(popt.fis);
	parameters.push_back(popb.fis);
	parameters.push_back(popTotal.af);
	parameters.push_back(0.1);
	parameters.push_back(popTotal.af);

	double sums [6] = {0};
	double fsts [10000]  ;

	for(int i = 0; i < 15000; i++){

	  // update each of j parameters

	  for(int j = 0; j < 6; j++ ){

	    updateParameters(popt, popb, parameters, j);
	    if(i > 4999){
	      sums[j]     += parameters[j];
	    }
	  }
	  if(i > 4999){
	    fsts[i - 5000] =  parameters[5];
	  }
	  for(vector<int>::iterator itt = popt.questionable.begin(); itt != popt.questionable.end(); itt++){
	    updateGenotypes(popt, popb, parameters, (*itt), 0);

	  }
	  for(vector<int>::iterator itb = popb.questionable.begin(); itb != popb.questionable.end(); itb++){
	    updateGenotypes(popt, popb, parameters, (*itb) , 1);
	  }
	}

	qsort (fsts, sizeof(fsts)/sizeof(fsts[0]), sizeof(fsts[0]), cmp );

	double lcredint = fsts[500];
	double hcredint = fsts[9500];

    	cout << var.sequenceName << "\t"  << var.position
	     << "\t"  << popt.af
             << "\t"  << sums[0]/10000
	     << "\t"  << popb.af
	     << "\t"  << sums[1]/10000
	     << "\t"  << popTotal.af
	     << "\t"  << sums[4]/10000
	     << "\t"  << sums[5]/10000
	     << "\t"  << lcredint
	     << "\t"  << hcredint
	     << endl;
    }
    return 0;
}
